### PR TITLE
`msg-sim` crate implementation

### LINKED ISSUE #53: Proposal: `msg-sim` crate
## Context
Being able to simulate latency, bandwidth, packet loss etc. ("shaping traffic") to reflect the real world is all incredibly useful for testing & benchmarking different protocols. It's how we discovered that QUIC in most cases is actually slower than TCP. We used [comcast](https://github.com/tylertreat/comcast) to add 50ms of latency and a max bandwidth of 1Gbps to localhost, and then ran our benchmarks again.

## Proposal
It would be nice to be able to shape traffic in unit tests, integration tests and benchmarks inside of MSG, so I'm proposing the `msg-sim` crate to do just that. `msg-sim` will have the following features:
- Support both MacOS and Linux
- Highly configurable
- Standalone crate for use in other projects
- Will be able to do everything comcast can do

## Simulation
We usually use localhost for testing, which is the loopback interface. Aside from it being a virtual network interface, it also has a much higher MTU than normal, doesn't use Ethernet etc.

### PR DESCRIPTION
**Pull request by mempirate**
Closes #53

### CODE REVIEW

**Comment by mempirate** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/chainbound/msg-rs/54?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #48** <a href="https://app.graphite.dev/github/pr/chainbound/msg-rs/48?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #54** <a href="https://app.graphite.dev/github/pr/chainbound/msg-rs/54?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/chainbound/msg-rs/54?utm_source=stack-comment).

**Review by merklefruit — APPROVED** [review]
LGTM

File: msg-sim/src/dummynet.rs [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+        // Build the pipe.
+        self.pipe.build()?;
+
+        self.create_loopback_alias()?;
+
+        self.load_pf_config()?;
+        self.apply_dummynet_rule()?;
+
+        // Enable the packet filter
+        let status = Command::new("sudo").args(["pfctl", "-E"]).status()?;
```
[the remark this answers is not in the document]
Comment (mempirate) [reply msg-sim/src/dummynet.rs:163]: Unfortunately yes, I disabled these tests in CI for now

**Comment by mempirate** [comment]
### Merge activity

* **Jan 18, 9:01 AM**: @mempirate started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/chainbound/msg-rs/54).

**Pull request merged**